bundle/Service/ScheduledVisibilityService.php

```diff
@@ -0,0 +1,129 @@
+     */
+    public function updateVisibilityIfNeeded(Content $content, ?string $handlerIdentifier = null): VisibilityUpdateResult
+    {
+        if (!$this->accept($content)) {
+            return VisibilityUpdateResult::NoChange;
+        }
+
+        $handler = $this->registry->get($handlerIdentifier);
+
+        if ($this->shouldBeHidden($content) && !$handler->isHidden($content)) {
+            $handler->hide($content);
+
+            return VisibilityUpdateResult::Hidden;
+        }
+
+        if ($this->shouldBeVisible($content) && $handler->isHidden($content)) {
+            $handler->reveal($content);
+
+            return VisibilityUpdateResult::Revealed;
+        }
+
+        return VisibilityUpdateResult::NoChange;
+    }
+
+    public function accept(Content $content): bool
+    {
+        $contentType = $content->getContentType();
+
+        $fieldDefinitions = $contentType->getFieldDefinitions();
+        if (!$fieldDefinitions->has('publish_from') || !$fieldDefinitions->has('publish_to')) {
+            return false;
+        }
+
+        /** @var Field $publishFromField */
+        $publishFromField = $content->getField('publish_from');
+
+        /** @var Field $publishToField */
+        $publishToField = $content->getField('publish_to');
+        if (
+            ($publishFromField->fieldTypeIdentifier !== 'ezdatetime' && $publishFromField->fieldTypeIdentifier !== 'ezdate')
+            || ($publishToField->fieldTypeIdentifier !== 'ezdatetime' && $publishToField->fieldTypeIdentifier !== 'ezdate')
+        ) {
+            return false;
+        }
+
+        $publishFromDateTime = $this->getDateTime($publishFromField);
+        $publishToDateTime = $this->getDateTime($publishToField);
+        if ($publishFromDateTime !== null && $publishToDateTime !== null && $publishToDateTime <= $publishFromDateTime) {
+            return false;
+        }
+
+        return true;
+    }
+
+    public function shouldBeHidden(Content $content): bool
+    {
+        $publishFromField = $content->getField('publish_from');
+        $publishToField = $content->getField('publish_to');
+
+        $publishFromDateTime = $this->getDateTime($publishFromField);
+        $publishToDateTime = $this->getDateTime($publishToField);
+
+        $currentDateTime = new DateTime();
+        if (($publishFromDateTime !== null && $publishFromDateTime > $currentDateTime)
+            || ($publishToDateTime !== null && $publishToDateTime <= $currentDateTime)) {
+            return true;
+        }
+
+        return false;
+    }
+
+    public function shouldBeVisible(Content $content): bool
+    {
+        $publishFromField = $content->getField('publish_from');
+        $publishToField = $content->getField('publish_to');
+
+        $publishFromDateTime = $this->getDateTime($publishFromField);
+        $publishToDateTime = $this->getDateTime($publishToField);
+
+        if ($publishToDateTime === null && $publishFromDateTime === null) {
+            return true;
```

Comment:
Just realized that by the old implementation we should perform no changes if the dates are not set. Hence this should actually return false here - for example it might be hidden intentionally by the editor. Similar is then the case for `shouldBeHidden()` - it should not be hidden if the dates are not set.

Reply:
fixed in cde1a7391eb2eb37a59170efd5fad9d64f3684bf